hold Matrix4 elements in double precision until uploaded to GL

foijord opened the pull request:
I'm working on a project that involves drawing the whole earth in an ECEF coordinate system. For this to work, we need double-precision model and view matrices. Currently, single precision matrices are used, causing loss of precision in camera movement and rendering artifacts (cracks between tiles). If the model and view matrices is stored in double precision, large values in these matrices tend to cancel out and the resulting modelview matrix can be safely converted to float32.

This pull request solves the problem for our project, but there are possible performance implications with the change, as it's creating a new Float32Array each frame. I'm looking for input on how to proceed with this change.

Commit:
hold Matrix4 elements in double precision until uploaded to GL

mrdoob commented on build/three.js:
```diff
@@ -4482,8 +4482,9 @@
-		gl.uniformMatrix4fv( this.addr, false, v.elements || v );
-
+		if ( v.elements === undefined ) { gl.uniformMatrix4fv( this.addr, false, v ); }
+		else { gl.uniformMatrix4fv( this.addr, false, new Float32Array(v.elements) ); }
```
So you're proposing creating a Float32Array every time an uniform gets uploaded?

WestLangley replied:
The OP admits in the post that is a hack and is looking for suggestions.

mrdoob commented:
Maybe it would be better to have a setting in the library that makes `Matrix4` use `Float64Array` instead of `Float32Array`?

WestLangley commented:
Perhaps the user can change the scale of world units from meters to kilometers, for example, to avoid large world-unit values.

Commit:
fix typo

foijord commented:
Thanks for the suggestions. 
I don't think using a Float64Array would solve anything, as it would need to be converted to a Float32Array before pushing to GL.

And changing scales doesn't accomplish anything. Going to kilometers would make the numbers 1000 times smaller in absolute size, but the accuracy requirements would increase by 1000x. Only double precision values in the elements will do the trick.

If I can get rid of the Float32Array creation on each upload (by creating a "static" float32 array that I would copy to before uploading?), would this then become an acceptable pull request? or is more work needed, e.g. make this consistent between all uniforms, or perhaps make it a feature you must turn on to enable. Then there's the question of something like this would break client apps (e.g. someone accessing the elements directly).

Just trying to get a handle on the amount of work involved and probability of getting a request like this accepted before putting in significant amount of work on this...

Commit:
avoid creating float32 array each uniform upload

WestLangley commented:
Can you provide a simple live example to demonstrate the issue? (Not your entire project.)

foijord commented:
Alright, I'll provide an example.

Commit:
hacked interactive cubes example to show floating point issues with float32 matrices

foijord commented:
I've hacked the 'webgl_interactive_cubes.html' example to show the issue. If you run this example now, you should see something like this: 
![image](https://cloud.githubusercontent.com/assets/1118749/22645695/94c45320-ec68-11e6-8ddd-e2ccad4e0816.png)
The cubes are laid out edge to edge, made smaller and placed at an offset from the origin.

Now, undo my double-precision Matrix4 fix: git checkout 6c88bb3 build/
and you should see something this:
![image](https://cloud.githubusercontent.com/assets/1118749/22645664/6eb5fed6-ec68-11e6-8a41-9ee2b5155df5.png)
Notice the cracks between some of the cubes, and that they overlap some places. This is due to rounding errors from the float32 values in the Matrix4. 

So if we keep the model and view matrices in double precision, the large values from the offset in the model matrix is canceled out by the large offset in the view matrix and the resulting model-view matrix doesn't contain large values any longer and can safely be converted to float32 and uploaded to the GPU.

This, together with the awesome logarithmic depth buffer feature, allows us to draw extremely large scenes, way beyond planetary-scale.

mrdoob commented:
https://rawgit.com/foijord/three.js/double-precision-uniforms/examples/webgl_interactive_cubes.html

mrdoob commented:
This actually sounds pretty good to me. I wonder if there are performance implications though.

mrdoob commented:
/cc @bhouston

bhouston commented:
This is a strange issue that will arise when you want precision away from the origin with fp32 matrices.  This issue is present in UE4 and Unity 3D and nearly every other game engine.

The cost to switching to doubles for Matrix4 is not worth it because it will slow down everyone else for a special case situation.

My suggestion would be either to create a special Matrix4D that you can convert to a Matrix4 - that is likely easy.  Maybe contribute it to ```examples/js```?

Or you can do transforms to bring the things you care about to near the origin when they are going to be displayed on screen: http://answers.unity3d.com/questions/54739/any-solution-for-extreamly-large-gameworlds-single.html

I guess I do not want to move ThreeJS towards double precision because of this issue, but you do not have to do what I suggest.

bhouston commented:
More details of UE4:

https://forums.unrealengine.com/showthread.php?109922-What-happened-to-the-quot-large-world-coordinate-system-quot-feature-the-highest-voted-in-the-roadmap

Although maybe a feature somewhere to switch ThreeJS to double precision for those that need it?  This way you get your precision while others do not have to pay for the speed penalty.  Double is a lot slower than Single.  I understand that many mobile phones are currently horribly slow with doubles - but I do not have hard numbers.

foijord commented:
I'm questioning the assumption this change would slow things down. All Matrix operations are in double-precision already, and so is probably almost all of THREE.js. Take matrix multiplication for example, where the values in the float32 array are copied into temporary double-precision numbers before they are multiplied/added up and then stored back in the 32 bit float array. This involves a conversion from float32 to float64 and back. My suggested change would simply get rid of this conversion, nothing else would change. Might even be faster!

bhouston commented:
> Take matrix multiplication for example, where the values in the float32 array are copied into temporary double-precision numbers before they are multiplied/added up and then stored back in the 32 bit float array.

We will have to convert to fp32 on upload.  I am okay to try this change.  I would love to see some benchmarks to show that this is actually faster.  If it is, I think that would be amazing.

I just checked and it seems that fp64 performance on mobile is actually incredibly good -- I figured it would be horrible, but I guess I am wrong?  http://www.anandtech.com/show/6971/exploring-the-floating-point-performance-of-modern-arm-processors

bhouston commented:
I do know that the matrix multiplies when rendering can be a significant cost in ThreeJS for complex scene graphs.  So if we can make matrix multiplication faster, that would be a great contribution in general.

foijord commented:
OK, let me know if I should try to test the performance of this, or make further changes before it can be accepted.

mrdoob commented:
I think we just have to meassure these two:

https://rawgit.com/mrdoob/three.js/dev/examples/webgl_performance.html
https://rawgit.com/foijord/three.js/double-precision-uniforms/examples/webgl_performance.html

Commit:
hacked performance test to be reproducible: don't use random positions and disable camera movement

foijord commented:
I'm surprised by the difference in fps you are seeing. On my end, there's no such difference. I'm on Windows 7, Chrome 56.0.2924.87 (64-bit), Quadro K6000, so completely CPU bound. I'm curious to know what you're running on.

I noticed that the perf example has varying fps based on camera position, and the monkeys are placed out randomly within a box, something that could perhaps make it less repetable. So i've removed the camera movement, and placed the monkeys on a regular grid to get more consistent results. I also added more monkeys to get the fps down from 60. I get about 55 fps on both of these. 

hacked version using vanilla three.js:
https://rawgit.com/foijord/three.js/dev/examples/webgl_performance.html

hacked version using double-precision fix for Matrix4
https://rawgit.com/foijord/three.js/double-precision-uniforms/examples/webgl_performance.html

What's your performance on these updated versions?

bhouston commented:
I profiled this quality on an Qualcomm 820 / Adreno 530 with an HD screen and I get 22fps on both versions.  Which is nice. :)

On iPad Air 2 I get 12fps with the old and 11fps the new double Matrix4 - which is pretty close.

My desktop runs at 60fps on both versions so I can not see a difference.

Commit:
add more monkeys

foijord commented:
Added more monkeys, I'm down to about 30 fps now, both versions.

bhouston commented:
On my desktop I am getting results where the double-precision-uniforms (which is not an accurate branch name, but whatever) is getting about 4fps more than the plain version (57fps versus 54fps.)  The test is pretty variable and changes its fps on refreshes of the page, but the double-precision-uniforms branch seems to be consistently faster.  I've refreshed about 8 times each and there is a difference here - I assume it isn't side effect of something else...

Commit:
attempt to increase performance by avoiding some copying

foijord commented:
I've updated the matrix multiplication method in the double-precision-uniforms branch. I think I got the fps up slightly by avoiding some copying, at the cost of code clarity. Perhaps this would help with performance on that iPad as well?

bhouston commented:
@foijord The technique you implemented to avoid copying would have also likely kept all the values as floats in the Float32Array branch, thus that would have been an effective performance optimization as well.  

I have tried that style of optimization elsewhere, keeping everything in an array and it generally works well.  I am not sure if it is because it avoids copies or that works because it allows for the compiler to leave everything as Float32 rather than converting up to Float64.

Commit:
bugfix: multiplyMatrices can be called with this matrix as a or b, so copy elements if that's the case

foijord commented:
I noticed just now that multiplyMatrices is called with 'this' as argument a couple of places (multiply and premultiply), so copying is needed in those cases. Added a fix to adress that, which may or may not negate any performance gains.

Commit:
should probably use === instead of == to check if this is passed in

mrdoob commented on build/three.module.js:
```diff
@@ -3396,7 +3396,7 @@ Matrix4.prototype = {
 	copy: function ( m ) {
 
-		this.elements.set( m.elements );
+		this.elements = m.elements.slice();
```
Wouldn't this create too much GC?

foijord replied:
Fixed by copying element-by-element in both places where slice() was used

mrdoob commented:
> @foijord The technique you implemented to avoid copying would have also likely kept all the values as floats in the Float32Array branch, thus that would have been an effective performance optimization as well.  

Aww... Too good to be true...

![screen shot 2017-02-08 at 23 26 48](https://cloud.githubusercontent.com/assets/97088/22773331/724a3a86-ee56-11e6-82e0-00d13d1cc69a.png)

![screen shot 2017-02-08 at 23 27 01](https://cloud.githubusercontent.com/assets/97088/22773338/7a0facba-ee56-11e6-998e-61992a07fc03.png)

bhouston commented:
@mrdoob  wrote: "Aww... Too good to be true..."

You do not have to copy the two source arrays as you did.  Just create one Float32Array temp for the result of the multiply and copy that into the result.  Thus you have one less Float32Array copy.  If the numbers hold it is faster than the many temporaries.

bhouston commented:
If I remove one copy, the total time of your second version is 271ms, which is quite a bit less than ~400ms of the top version.

mrdoob commented:
![screen shot 2017-02-08 at 23 52 49](https://cloud.githubusercontent.com/assets/97088/22773982/c4e99478-ee59-11e6-9b18-311bd7b9a676.png)

Yeah, caching seems to still be faster.

mrdoob commented:
Hmm... 

![screen shot 2017-02-08 at 23 58 05](https://cloud.githubusercontent.com/assets/97088/22774160/ae591138-ee5a-11e6-93c6-e37ef5eb982d.png)

@foijord do you mind trying this?

```js
multiplyMatrices: ( function () {

	var te = [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ];

	return function ( a, b ) {

		var ae = a.elements;
		var be = b.elements;

		te[ 0 ] = ae[ 0 ] * be[ 0 ] + ae[ 4 ] * be[ 1 ] + ae[ 8 ] * be[ 2 ] + ae[ 12 ] * be[ 3 ];
		te[ 4 ] = ae[ 0 ] * be[ 4 ] + ae[ 4 ] * be[ 5 ] + ae[ 8 ] * be[ 6 ] + ae[ 12 ] * be[ 7 ];
		te[ 8 ] = ae[ 0 ] * be[ 8 ] + ae[ 4 ] * be[ 9 ] + ae[ 8 ] * be[ 10 ] + ae[ 12 ] * be[ 11 ];
		te[ 12 ] = ae[ 0 ] * be[ 12 ] + ae[ 4 ] * be[ 13 ] + ae[ 8 ] * be[ 14 ] + ae[ 12 ] * be[ 15 ];

		te[ 1 ] = ae[ 1 ] * be[ 0 ] + ae[ 5 ] * be[ 1 ] + ae[ 9 ] * be[ 2 ] + ae[ 13 ] * be[ 3 ];
		te[ 5 ] = ae[ 1 ] * be[ 4 ] + ae[ 5 ] * be[ 5 ] + ae[ 9 ] * be[ 6 ] + ae[ 13 ] * be[ 7 ];
		te[ 9 ] = ae[ 1 ] * be[ 8 ] + ae[ 5 ] * be[ 9 ] + ae[ 9 ] * be[ 10 ] + ae[ 13 ] * be[ 11 ];
		te[ 13 ] = ae[ 1 ] * be[ 12 ] + ae[ 5 ] * be[ 13 ] + ae[ 9 ] * be[ 14 ] + ae[ 13 ] * be[ 15 ];

		te[ 2 ] = ae[ 2 ] * be[ 0 ] + ae[ 6 ] * be[ 1 ] + ae[ 10 ] * be[ 2 ] + ae[ 14 ] * be[ 3 ];
		te[ 6 ] = ae[ 2 ] * be[ 4 ] + ae[ 6 ] * be[ 5 ] + ae[ 10 ] * be[ 6 ] + ae[ 14 ] * be[ 7 ];
		te[ 10 ] = ae[ 2 ] * be[ 8 ] + ae[ 6 ] * be[ 9 ] + ae[ 10 ] * be[ 10 ] + ae[ 14 ] * be[ 11 ];
		te[ 14 ] = ae[ 2 ] * be[ 12 ] + ae[ 6 ] * be[ 13 ] + ae[ 10 ] * be[ 14 ] + ae[ 14 ] * be[ 15 ];

		te[ 3 ] = ae[ 3 ] * be[ 0 ] + ae[ 7 ] * be[ 1 ] + ae[ 11 ] * be[ 2 ] + ae[ 15 ] * be[ 3 ];
		te[ 7 ] = ae[ 3 ] * be[ 4 ] + ae[ 7 ] * be[ 5 ] + ae[ 11 ] * be[ 6 ] + ae[ 15 ] * be[ 7 ];
		te[ 11 ] = ae[ 3 ] * be[ 8 ] + ae[ 7 ] * be[ 9 ] + ae[ 11 ] * be[ 10 ] + ae[ 15 ] * be[ 11 ];
		te[ 15 ] = ae[ 3 ] * be[ 12 ] + ae[ 7 ] * be[ 13 ] + ae[ 11 ] * be[ 14 ] + ae[ 15 ] * be[ 15 ];

		this.elements = te.slice();

		return this;

	};

} )(),
```

mrdoob commented:
Or maybe...

```js
multiplyMatrices: ( function () {

	var e = [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ];

	return function ( a, b ) {

		var ae = a.elements;
		var be = b.elements;
		var te = this.elements;

		e[ 0 ] = ae[ 0 ] * be[ 0 ] + ae[ 4 ] * be[ 1 ] + ae[ 8 ] * be[ 2 ] + ae[ 12 ] * be[ 3 ];
		e[ 4 ] = ae[ 0 ] * be[ 4 ] + ae[ 4 ] * be[ 5 ] + ae[ 8 ] * be[ 6 ] + ae[ 12 ] * be[ 7 ];
		e[ 8 ] = ae[ 0 ] * be[ 8 ] + ae[ 4 ] * be[ 9 ] + ae[ 8 ] * be[ 10 ] + ae[ 12 ] * be[ 11 ];
		e[ 12 ] = ae[ 0 ] * be[ 12 ] + ae[ 4 ] * be[ 13 ] + ae[ 8 ] * be[ 14 ] + ae[ 12 ] * be[ 15 ];

		e[ 1 ] = ae[ 1 ] * be[ 0 ] + ae[ 5 ] * be[ 1 ] + ae[ 9 ] * be[ 2 ] + ae[ 13 ] * be[ 3 ];
		e[ 5 ] = ae[ 1 ] * be[ 4 ] + ae[ 5 ] * be[ 5 ] + ae[ 9 ] * be[ 6 ] + ae[ 13 ] * be[ 7 ];
		e[ 9 ] = ae[ 1 ] * be[ 8 ] + ae[ 5 ] * be[ 9 ] + ae[ 9 ] * be[ 10 ] + ae[ 13 ] * be[ 11 ];
		e[ 13 ] = ae[ 1 ] * be[ 12 ] + ae[ 5 ] * be[ 13 ] + ae[ 9 ] * be[ 14 ] + ae[ 13 ] * be[ 15 ];

		e[ 2 ] = ae[ 2 ] * be[ 0 ] + ae[ 6 ] * be[ 1 ] + ae[ 10 ] * be[ 2 ] + ae[ 14 ] * be[ 3 ];
		e[ 6 ] = ae[ 2 ] * be[ 4 ] + ae[ 6 ] * be[ 5 ] + ae[ 10 ] * be[ 6 ] + ae[ 14 ] * be[ 7 ];
		e[ 10 ] = ae[ 2 ] * be[ 8 ] + ae[ 6 ] * be[ 9 ] + ae[ 10 ] * be[ 10 ] + ae[ 14 ] * be[ 11 ];
		e[ 14 ] = ae[ 2 ] * be[ 12 ] + ae[ 6 ] * be[ 13 ] + ae[ 10 ] * be[ 14 ] + ae[ 14 ] * be[ 15 ];

		e[ 3 ] = ae[ 3 ] * be[ 0 ] + ae[ 7 ] * be[ 1 ] + ae[ 11 ] * be[ 2 ] + ae[ 15 ] * be[ 3 ];
		e[ 7 ] = ae[ 3 ] * be[ 4 ] + ae[ 7 ] * be[ 5 ] + ae[ 11 ] * be[ 6 ] + ae[ 15 ] * be[ 7 ];
		e[ 11 ] = ae[ 3 ] * be[ 8 ] + ae[ 7 ] * be[ 9 ] + ae[ 11 ] * be[ 10 ] + ae[ 15 ] * be[ 11 ];
		e[ 15 ] = ae[ 3 ] * be[ 12 ] + ae[ 7 ] * be[ 13 ] + ae[ 11 ] * be[ 14 ] + ae[ 15 ] * be[ 15 ];

		for ( var i = 0; i < 16; i ++ ) te[ i ] = e[ i ];

		return this;

	};

} )(),

mrdoob commented:
But, I bet the original `multiplyMatrices` works the best for the `Array` approach.

mrdoob commented on build/three.module.js:
```diff
@@ -4764,7 +4768,7 @@ function WebGLUniforms( gl, program, renderer ) {
 	var n = gl.getProgramParameter( program, gl.ACTIVE_UNIFORMS );
 
-	for ( var i = 0; i !== n; ++ i ) {
+	for ( var i = 0; i < n; ++ i ) {
```
Do you mind reverting this change?

Commit:
improved matrix multiply by mrdoob

Commit:
remove accidentally committed file

Commit:
remove some GC pressure by copying element by element

mrdoob commented:
![screen shot 2017-02-09 at 01 01 50](https://cloud.githubusercontent.com/assets/97088/22776165/80843ffe-ee63-11e6-97ea-ecfed1ca905d.png)

mrdoob commented:
Anyway...

> hacked version using vanilla three.js:
> https://rawgit.com/foijord/three.js/dev/examples/webgl_performance.html
>
> hacked version using double-precision fix for Matrix4
> https://rawgit.com/foijord/three.js/double-precision-uniforms/examples/webgl_performance.html

double-precision seems to be slightly faster now on my system.

Commit:
revert unwanted change

mrdoob commented:
I'll merge for now! 👍

mrdoob commented:
Thanks!

foijord commented:
Cool, thanks! This is a really awesome change for our project :)

WestLangley commented:
Do you intend to do the same for `Matrix3`?

mrdoob commented:
Eek... It seems that some code already relies on `matrix.elements` having a `.set()` method 😕

https://rawgit.com/mrdoob/three.js/dev/examples/#webgl_loader_sea3d

foijord commented:
I haven't thought about it, but I don't mind doing the same for Matrix3. Shouldn't take long now that we know how...

foijord commented:
I guess client code shouldn't access the elements array directly anyway. Should just use the copy method here.

foijord commented:
This went a little fast I think, I never meant to merge my hacks in the example code. That was just to get some repeatable results to judge the performance on. 

Also, since we broke the sea3d examples, would you like me to fix that and create another pull request?

Please advise...

bhouston commented:
@foijord Please do a another PR that fixes Sea3D and cleans up any hacks you think are in the code base that shouldn't be. :)  I think you and @mrdoob collaborated great on this change.  More precise and faster!

bhouston commented:
Also can you update Matrix3 with the same fast techniques and doubles?  @foijord   Also please check every example to see if they still load without errors.  It is manually but that is what I do to check major core changes to see if they break things.

foijord commented:
The pull requests for cleaning up the hacks and fixes the Sea3D stuff is already made. I can implement this for Matrix3D as well just for consistency, but it's probably not needed, since a matrix3 will not represent translations. But it's a good idea to check all examples, will do that.

mrdoob commented:
@s3ththompson @bmeurer this is another issue where we would really benefit from some V8 guys input.

1. We currently use Float32Array for matrix operations. What do you think of the idea of using Javascript numbers for matrix computations and pass them to a float32array at gpu upload time? V8 seems to have similar performance and we gain a lot of precision.

2. How do you guys suggest we copy an `Array`?

```js
// A
for ( var i = 0; i < 16; i ++ ) array2[ i ] = array1[ i ];
// B
array2 = array1.slice(); // probably leaky?
```

mrdoob commented:
I wonder if we should rename  `elements` to `array` to avoid breakage...

```js
function Matrix4() {
	this.array = [
		1, 0, 0, 0,
		0, 1, 0, 0,
		0, 0, 1, 0,
		0, 0, 0, 1
	];
}

Object.defineProperty( Matrix4.prototype, 'elements', {
	get: function () {
		console.warn( 'Matrix4: elements has been renamed to array.' ); 
		return new Float32Array( this.array );
	}
} );
```

Or maybe make it "private" and rename it to `_elements`?

bmeurer commented:
My gut feeling is that option A is better for now. I think `Array.prototype.slice` is not optimized for that use case currently.

mrdoob commented:
@bmeurer Thanks!

mrdoob commented:
> I wonder if we should rename  `elements` to `array` to avoid breakage...

After pondering about it for a bit, I think it's not really worth it to try to avoid breakage in this case as it would result in a considerable slow down to code accessing the array directly anyway.

kenrussell commented:
Would it be possible to make a smaller benchmark comparing the typed array vs. Array approaches?

https://rawgit.com/mrdoob/three.js/dev/examples/webgl_performance.html
https://rawgit.com/foijord/three.js/double-precision-uniforms/examples/webgl_performance.html

The second link 404s now.

We should push for the bottlenecks slowing down Typed Arrays vs. regular Arrays to be fixed. If the benchmark could be isolated to not even need WebGL, that would be even better.

Thanks.
